gstd/src/common/errors.rs

```diff
@@ -33,50 +34,77 @@ pub type Result<T, E = Error> = core::result::Result<T, E>;
 /// Common error type returned by API functions from other modules.
 #[derive(Clone, Debug, Eq, PartialEq)]
 pub enum Error {
+    /* Protocol under-hood errors */
     /// [`gcore::errors::Error`] type.
+    ///
+    /// NOTE: this error could only be returned from syscalls.
     Core(CoreError),
-    /// Timeout reached while expecting for reply.
-    Timeout(u32, u32),
+
+    /* API lib under-hood errors */
     /// Conversion error.
+    ///
+    /// NOTE: this error returns from incorrect bytes conversion.
     Convert(&'static str),
-    /// Decoding error.
+
+    /// `scale-codec` decoding error.
+    ///
+    /// NOTE: this error returns from APIs that return specific `Decode` types.
     Decode(scale_info::scale::Error),
-    /// Reply code returned by another program.
-    ReplyCode(ReplyCode),
-    /// This error occurs when providing zero duration to waiting functions
-    /// (e.g. see `exactly` and `up_to` functions in
-    /// [CodecMessageFuture](crate::msg::CodecMessageFuture)).
-    EmptyWaitDuration,
-    /// This error occurs when providing zero gas amount to system gas reserving
-    /// function (see
-    /// [Config::set_system_reserve](crate::Config::set_system_reserve)).
-    ZeroSystemReservationAmount,
-    /// This error occurs when providing zero duration to mutex lock function
-    ZeroMxLockDuration,
+
+    /// Gstd API usage error.
+    ///
+    /// Note: this error returns from `gstd` APIs in case of invalid arguments.
+    Gstd(UsageError),
+
+    /* Business logic errors */
+    /// Received error reply while awaited response from another actor.
+    ///
+    /// NOTE: this error could only be returned from async messaging.
+    ErrorReply(String, ErrorReplyReason),
+
+    // TODO: consider should we meant unsupported replies as success.
```

Comment:
No TODOs without an issue

Reply:
that's the thing I was about to discuss on review: compatibility here is depend on "status" codes we use. I assume, that error codes will change a lot in future, while success ones are kinda constant and unlikely to be changed, so I think it's incorrect to mean unsupported replies as success ones. Wanna to hear other opinions

Reply:
@ark0f could you pls share your thoughts here ?